src/main/java/io/penguinstats/dao/ItemDropDaoCustomImpl.java

```diff
@@ -3,30 +3,139 @@
 import org.springframework.data.mongodb.core.query.Criteria;
-import org.springframework.data.mongodb.core.query.Query;
 
 import io.penguinstats.constant.AggregationOperationConstants;
+import io.penguinstats.enums.Server;
 import io.penguinstats.model.ItemDrop;
+import io.penguinstats.model.QueryConditions;
+import io.penguinstats.model.QueryConditions.StageWithTimeRange;
 
 public class ItemDropDaoCustomImpl implements ItemDropDaoCustom {
 
+	private static Logger logger = LogManager.getLogger(ItemDropDaoCustomImpl.class);
+
 	@Autowired
 	MongoTemplate mongoTemplate;
 	@Autowired
 	AggregationOperationConstants aggregationOperationConstants;
 
+	@Override
+	public List<Document> aggregateItemDropQuantities(QueryConditions conditions) {
+		Long currentTime = System.currentTimeMillis();
+
+		List<String> userIDs = conditions.getUserIDs();
+		List<String> itemIds = conditions.getItemIds();
+		List<Server> servers = conditions.getServers();
+		List<StageWithTimeRange> stages = conditions.getStages();
+		Long interval = conditions.getInterval();
+
+		List<AggregationOperation> operations = new LinkedList<>();
+
+		// Pipe 1: filter by isReliable, isDeleted, stageId and timestamp
+		List<Criteria> criteriasInAndInPipe1 = new ArrayList<>();
+
+		criteriasInAndInPipe1.add(Criteria.where("isReliable").is(true));
+
+		if (userIDs.isEmpty())
+			criteriasInAndInPipe1.add(Criteria.where("isDeleted").is(false));
+		else
+			criteriasInAndInPipe1.add(Criteria.where("userID").in(userIDs));
+
+		if (!servers.isEmpty())
+			criteriasInAndInPipe1.add(Criteria.where("server").in(servers));
+
+		if (!stages.isEmpty()) {
+			if (1 == stages.size() && stages.get(0).getStageId() == null) {
+				StageWithTimeRange stage = stages.get(0);
+				Long min = stage.getStart() == null ? 0L : stage.getStart();
+				Long max = stage.getEnd() == null ? System.currentTimeMillis() : stage.getEnd();
+				criteriasInAndInPipe1.add(Criteria.where("timestamp").gte(min).lt(max));
+			} else {
+				List<Criteria> criteriasInOrInPipe1 = new ArrayList<>();
+				stages.forEach(stage -> {
+					Long min = stage.getStart() == null ? 0L : stage.getStart();
+					Long max = stage.getEnd() == null ? System.currentTimeMillis() : stage.getEnd();
+					criteriasInOrInPipe1.add(new Criteria().andOperator(Criteria.where("timestamp").gte(min).lt(max),
+							Criteria.where("stageId").is(stage.getStageId())));
+				});
+				criteriasInAndInPipe1.add(new Criteria().orOperator(criteriasInOrInPipe1.toArray(new Criteria[0])));
+			}
+		}
+
+		operations.add(Aggregation.match(new Criteria().andOperator(criteriasInAndInPipe1.toArray(new Criteria[0]))));
+
+		// Pipe 2: project section number
+		if (interval != null) {
+			Long baseTime = null;
+			if (stages.isEmpty())
+				baseTime = 0L;
+			else {
+				final Long firstStartTime = stages.get(0).getStart();
+				boolean passCheck = true;
+				for (int i = 1, l = stages.size(); i < l; i++) {
```

Comment:
suggest to use other letters, as l looks similar to 1...

Reply:
fixed